content/en/docs/concepts/storage/persistent-volumes.md

```diff
@@ -710,6 +710,27 @@ spec:
       storage: 10Gi
 ```
 
+## Data Population
+
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+Data population is an experimental feature to allow persistent volumes to be created non-empty,
+from sources other than volumes or snapshots.
+
+To enable support for data population,  enable the `AnyVolumeDataSource` feature gate on the
+apiserver.
+
+This feature allows specifying any custom resource as the `DataSource` of a new PVC, as long
+as the custom resource is in the same namespace, and there is a controller installed to handle
+data population from that custom resource type (CRD). It's important to note that ordinary
+dynamic provisioners will ignore PVCs with `DataSource` types they don't support, so users
+need to be aware of what custom resources can be used for data population, and what
+controllers are installed.
+
+Users may create their own data populator CRDs and write their own data populator controllers
+to handle them. Such data populators are responsible for creating the persistent volume and
+binding it to the PVC on their own.
```

Comment:
This can be easily messed up and has been the cause of some security vulnerabilities in the past. Do we intend to have an example that people can follow?

Reply:
I have the hello-populator example from the original KEP. The plan was to present a formal design for how this should work and to get that approved as part of graduation to beta.

While writing these docs I realized how particularly unhelpful it is to have a feature gate with so little to say about it. Before beta we also need to address the issue of how users can find out what populators exist, and making sure to return good error messages when someone attempts to use something that's not installed.

Do you think it's better to say less? I don't want to encourage people to write code that will cause security problems.

Reply:
Yeah maybe until we have agreed upon pattern for populators we should not add a section here. Instead just document the feature gate.

Reply:
It removed this section and instead just documented the feature gate.